ssh-sk-client.c

```diff
@@ -43,15 +43,117 @@
 
 /* #define DEBUG_SK 1 */
 
+#ifdef WINDOWS
+static char module_path[PATH_MAX + 1];
+
+static char *
+find_helper_in_module_path(void)
+{
+	DWORD n;
+	char *ep;
+
+	memset(module_path, 0, sizeof(module_path));
+	if ((n = GetModuleFileNameA(NULL, module_path,
+	    sizeof(module_path) - 1)) == 0 || n >= sizeof(module_path) - 1) {
+		error_f("GetModuleFileNameA failed");
+		return NULL;
+	}
+	if ((ep = strrchr(module_path, '\\')) == NULL) {
+		error_f("couldn't locate trailing \\");
+		return NULL;
+	}
+	*(++ep) = '\0'; /* trim */
+	strlcat(module_path, "ssh-sk-helper.exe", sizeof(module_path) - 1);
+
+	return module_path;
+}
+
+static char *
+find_helper(void)
+{
+	char *helper;
+	char module_path[PATH_MAX + 1];
+	char *ep;
+	DWORD n;
+
+	if ((helper = getenv("SSH_SK_HELPER")) == NULL || strlen(helper) == 0) {
+		if ((helper = find_helper_in_module_path()) == NULL)
+			helper = _PATH_SSH_SK_HELPER;
+	}
+	if (!path_absolute(helper)) {
+		error_f("helper \"%s\" unusable: path not absolute", helper);
+		return NULL;
+	}
+	debug_f("using \"%s\" as helper", helper);
+
+	return helper;
+}
+
 static int
 start_helper(int *fdp, pid_t *pidp, void (**osigchldp)(int))
```

Comment:
Instead of duplicating this function between windows and Linux, can you have one function and have windows specific changes in #ifdef WINDOWS block. The advantage of this approach is any changes made to this function from upstream repository (unix repository) gets reflected on the windows side as well

Reply:
Addressed in dee6c5e7. In addition to my earlier comment on potential upstream preferences, the resulting sea of `#ifdef` blocks is a great place for subtle bugs to lurk, whereas a single block of code tends to be easier to reason about.